### ISSUE TITLE
Running much slower than standard TensorFlow 2.0

### ISSUE THREAD

**Issue by iwasnothing** [apple/tensorflow_macos#31]
I just compared the performance of the TensorFlow_macos package with the standard TensorFlow 2.0 python package.  I used a very simple tensor testing program, running on my Mac 2020 (x86), with the 2 package respectively.  I found that the tensorflow_macos is much slower.  I'm not sure whether the Mac-optimized package is only good for the new M1 ARM Mac
or it is the problem of my test program.  Here is the result and my code:

standard TensorFlow: 10.87s
tensorflow_macos: 79.41s

Code:
from datetime import datetime
import numpy as np
import tensorflow as tf
# Import mlcompute module to use the optional set_mlc_device API for device selection with ML Compute.
from tensorflow.python.compiler.mlcompute import mlcompute

# Select CPU device.
mlcompute.set_mlc_device(device_name="any") # Available options are 'cpu', 'gpu', and ‘any'.
print("start" , datetime.now())
X_raw = np.array([2013, 2014, 2015, 2016, 2017, 2018], dtype=np.float32)
y_raw = np.array([12000, 14000, 15000, 16500, 17500, 19000], dtype=np.float32)

X = (X_raw - X_raw.min()) / (X_raw.max() - X_raw.min())
y = (y_raw - y_raw.min()) / (y_raw.max() - y_raw.min())

X = tf.constant(X)
y = tf.constant(y)

a = tf.Variable(initial_value=0.)
b = tf.Variable(initial_value=0.)
variables = [a, b]

num_epoch = 10000
optimizer = tf.keras.optimizers.SGD(learning_rate=1e-3)
for e in range(num_epoch):
    # Use tf.GradientTape() to record information about the gradient of the loss function.
    with tf.GradientTape() as tape:
        y_pred = a * X + b
        loss = 0.5 * tf.reduce_sum(tf.square(y_pred - y))
    # TensorFlow computes the gradients of the loss function with respect to independent variables (model parameters) automatically.
    grads = tape.gradient(loss, variables)
    # TensorFlow updates parameters according to the gradient automatically.
    optimizer.apply_gradients(grads_and_vars=zip(grads, variables))

print(a, b)
print("end" , datetime.now())

**Comment by dkgaraujo** [comment]
Hi! I believe the problem with performance in your toy code is that it does not take advantage of the GPU parallelization. While you are running a massive number of epochs, the matrices you are multiplying are quite small in size, and therefore the overhead computing cost of transferring the data from the CPU to the GPU is taking a long time. Now, if both codes are being run on the CPU, it could be happening because some CPU math libraries are not yet incorporated in the tensorflow2.4 compared to the latest release of 2.3.

**Comment by iwasnothing** [comment]
Thanks @dkgaraujo , what kind of computation / model can take the advantage of it?  I just tried a model with Conv1D with LSTM, but it still taking much slower.  Standard tensorflow took about 40ms per step, but tensorflow_macos took over 400ms, which is 10X slower.

Here is my model code:
    model = tf.keras.models.Sequential([
      tf.keras.layers.Conv1D(filters=window_size, kernel_size=5,
                          strides=1, padding="causal",
                          activation="relu",
                          input_shape=[None, 1]),
      tf.keras.layers.LSTM(window_size, return_sequences=True),
      tf.keras.layers.LSTM(window_size, return_sequences=True),
      tf.keras.layers.Dense(window_size, activation="relu"),
      tf.keras.layers.Dense(10, activation="relu"),
      tf.keras.layers.Dense(1)

    ])

**Comment by TortoiseHam** [comment]
it needs to execute in non-eager mode I think, so wrapping your loop in a @tf.function would be required. You could also use model.fit() instead of the custom loop, though it seems to not realize that it needs to use the gpu by default right now. I got a 6x speedup on the following:


import tensorflow as tf
from tensorflow.python.keras import layers
from tensorflow.python.compiler.mlcompute import mlcompute

cifar10 = tf.keras.datasets.cifar10

def residual(x, num_channel):
    x = layers.Conv2D(num_channel, 3, padding='same')(x)
    x = layers.BatchNormalization(momentum=0.8)(x)
    x = layers.LeakyReLU(alpha=0.1)(x)
    x = layers.Conv2D(num_channel, 3, padding='same')(x)
    x = layers.BatchNormalization(momentum=0.8)(x)
    x = layers.LeakyReLU(alpha=0.1)(x)
    return x


def my_model():
    # prep layers
    inp = layers.Input(shape=(32, 32, 3))
    x = layers.Conv2D(64, 3, padding='same')(inp)
    x = layers.BatchNormalization(momentum=0.8)(x)
    x = layers.LeakyReLU(alpha=0.1)(x)
    # layer1
    x = layers.Conv2D(128, 3, padding='same')(x)
    x = layers.MaxPool2D()(x)
    x = layers.BatchNormalization(momentum=0.8)(x)
    x = layers.LeakyReLU(alpha=0.1)(x)
    x = layers.Add()([x, residual(x, 128)])
    # layer2
    x = layers.Conv2D(256, 3, padding='same')(x)
    x = layers.MaxPool2D()(x)
    x = layers.BatchNormalization(momentum=0.8)(x)
    x = layers.LeakyReLU(alpha=0.1)(x)
    # layer3
    x = layers.Conv2D(512, 3, padding='same')(x)
    x = layers.MaxPool2D()(x)
    x = layers.BatchNormalization(momentum=0.8)(x)
    x = layers.LeakyReLU(alpha=0.1)(x)
    x = layers.Add()([x, residual(x, 512)])
    # layers4
    x = layers.GlobalMaxPool2D()(x)
    x = layers.Flatten()(x)
    x = layers.Dense(10)(x)
    x = layers.Activation('softmax', dtype='float32')(x)
    model = tf.keras.Model(inputs=inp, outputs=x)

    return model

def my_data():
	(x_train, y_train),(x_test, y_test) = cifar10.load_data()
	x_train, x_test = x_train / 255.0, x_test / 255.0
	return x_train, y_train, x_test, y_test

if __name__ == "__main__":
    x_train, y_train, x_test, y_test = my_data()
    mlcompute.set_mlc_device(device_name='gpu')
    model = my_model()
    model.compile(optimizer='adam',
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])
    model.fit(x_train, y_train, epochs=10)
    model.evaluate(x_test, y_test)

**Comment by iwasnothing** [comment]
Thanks @TortoiseHam , I tried your program, it really has 8X improvement.  My LSTM program also use model.fit.  I think it is because I use TensorFlow.Dataset to load the data, that may not be suitable for loading to GPU.

**Comment by TortoiseHam** [comment]
@iwasnothing , maybe try replacing

mlcompute.set_mlc_device(device_name="any")

with 

mlcompute.set_mlc_device(device_name="gpu")

I found that leaving the framework to decide meant that on model.fit it would use the cpu for some reason

**Comment by popfido** [comment]
it's quite interesting that I got nearly the same performance when using `device_name="any"` compared with `device_name="gpu"`. So the framework does recognize the existence of my gpu here. Interesting.

The test code offered by @TortoiseHam running on 2019 Macbook Pro with Radeon Pro 5600M gets 10X faster performance compared with official version of tensorflow. Under CPU mode it also gets about 1.25X faster.
